File: leetcode/string/java/1946_largest_number/Solution.java
class Solution {
  public String maximumNumber(String num, int[] change) {
       
   
        //find i where we start to mutate
        int start = -1;
        int ct = 0;
        for (int i=0; i<num.length(); i++) {
            ct = num.charAt(i) - '0';
            
            if (change[ct] >ct) {
                start = i;
                break;
            }
        }
        
        //do not find and we are done
        if (start==-1) {
            return num;
        }
        
        //let us change from star and find the end
        int end = -1;
        
        for (int i=start; i<num.length(); i++) {
           
            ct = num.charAt(i) - '0';
            
            if (change[ct] <ct) {
                end = i-1;
                break;
            }
            
        }
        
        //if we do not find, it goes to the end
        if (end == -1) {
            end = num.length()-1;
        }
        
    
     
       //Using String Builder to ensure linear time complexity as java strings are immutable
       //for this question. running time from 865 to 14ms!!
       StringBuilder result = new StringBuilder("");
       
       for (int i=start; i<=end; i++) {
          
           ct = num.charAt(i) - '0';
           result.append(change[ct]);
       }
       
       return num.substring(0, start) + result.toString() + num.substring(end+1, num.length());
        
     
   }
}

File: leetcode/string/php/1946_Largest_Number/Solution.php
class Solution {

    /**
     * @param String $num
     * @param Integer[] $change
     * @return String
     */
    function maximumNumber($num, $change) {
        //find start
        $start = -1;
        
        for($i=0; $i<strlen($num); $i++) {
            if ($change[$num[$i]]  > $num[$i]) {
                $start = $i;
                break;
            }
        }
        
   
        if (-1 == $start) {
            return $num;
        }
        
        //find end
        $end = -1;
        
         for($i=$start+1; $i<strlen($num); $i++) {
            if ($change[$num[$i]]  < $num[$i]) {
                $end = $i-1;
                break;
            }
        }
        
        if ($end == -1) {
            $end = strlen($num) -1;
        }
        
     
        
        //build mid
        $mid = "";
        
        for ($i=$start; $i<=$end; $i++) {
            $mid =  $mid . $change[$num[$i]];
        }
        
        return substr($num, 0, $start) . $mid . substr($num, $end+1);
        
        
        
    }
}
